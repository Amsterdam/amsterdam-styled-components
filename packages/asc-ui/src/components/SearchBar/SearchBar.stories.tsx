import * as React from 'react'
import { storiesOf } from '@storybook/react/'
import { action } from '@storybook/addon-actions'
import SearchBar, { SearchBarStyle, IconButtonStyle } from './SearchBar'
import styled from '../../styled-components'
import { svgFill } from '../../styles/utils'

const SearchBarStory: React.FC<{}> = () => (
  <SearchBar
    placeholder="Enter the search text"
    onTextChanged={(text: string) => {
      console.log('TCL: text', text)

      action(`text changed: ${text}`)
    }}
    onSearch={(text: string) => {
      console.log('TCL: text', text)
      action(`button clicked: ${text}`)
    }}
  />
)

const SearchBarStoryWithChangedStyle: React.FC<{}> = () => {
  const ChangedSearchBarStyle = styled(SearchBarStyle)`
    align-content: flex-start;

    & > ${IconButtonStyle} {
      order: -1;
      margin-left: 0px;
      margin-right: 5px;
      padding: 10px;

      & svg {
        ${svgFill('tint', 'level5')};
      }
    }
  `

  return (
    <SearchBar
      styledComponent={ChangedSearchBarStyle}
      onTextChanged={(text: string) => action(`text changed: ${text}`)}
      onSearch={(text: string) => action(`button clicked: ${text}`)}
    />
  )
}

storiesOf('Composed/SearchBar', module)
  .add('default state', () => <SearchBarStory />)
  .add('with changed style', () => <SearchBarStoryWithChangedStyle />)
