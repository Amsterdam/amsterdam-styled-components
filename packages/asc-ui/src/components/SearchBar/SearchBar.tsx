// import { Search } from '@amsterdam/asc-assets'
import {
  ChangeEvent,
  FormEvent,
  HTMLAttributes,
  // RefObject,
  useEffect,
  useState,
  forwardRef,
} from 'react'
import { deprecatedWarning } from '../../utils'
import Button from '../Button'
import Icon from '../Icon'
import { InputProps } from '../Input'
import InputContext from '../Input/InputMethodsContext'
import TextField, { TextFieldProps } from '../TextField/TextField'
import SearchBarStyle, { Props as SearchBarStyleProps } from './SearchBarStyle'

export interface SearchBarProps extends TextFieldProps, SearchBarStyleProps {
  label?: string
  inputProps?: InputProps
  /**
   * @deprecated, use onChange in combo with onClear instead
   * @param value
   */
  onWatchValue?: (value: string) => void
  /**
   * @deprecated, wrap this component inside a <form onSubmit={() => ...}> instead
   * @param value
   */
  onSubmit?: (e: FormEvent) => void
  autoFocus?: boolean
}

const SearchBar = forwardRef<
  HTMLInputElement,
  SearchBarProps & HTMLAttributes<HTMLInputElement>
>(
  (
    {
      children,
      placeholder,
      onSubmit,
      onBlur,
      onChange,
      onFocus,
      onWatchValue,
      value,
      label,
      hideAt,
      showAt,
      onClear,
      inputProps,
      autoFocus,
      ...otherProps
    },
    inputRef,
  ) => {
    // let inputRef: RefObject<HTMLInputElement> | null = null
    const [inputValue, setInputValue] = useState(value || '')

    console.log('search', inputRef)

    const handelOnClear = () => {
      setInputValue('')
      if (inputRef && inputRef.current) {
        inputRef.current.focus()
      }

      if (onClear) {
        onClear()
      }
    }

    const handleOnSubmit = (e: FormEvent) => {
      if (onSubmit) {
        deprecatedWarning(
          `onSubmit is about to be deprecated, wrap this component inside a <form onSubmit={...} /> instead`,
        )
        onSubmit(e)
      }
    }

    const handleOnChange = (e: ChangeEvent<HTMLInputElement>) => {
      setInputValue(e.currentTarget.value)

      if (onChange) {
        onChange(e)
      }
    }

    useEffect(() => {
      if (onWatchValue) {
        deprecatedWarning(
          'onWatchValue is about to be deprecated. Use onChange in combo with onClear instead',
        )
        // @ts-ignore
        onWatchValue(inputValue)
      }
    }, [inputValue, onWatchValue])

    useEffect(() => {
      if (typeof value !== 'undefined') {
        setInputValue(value)
      }
    }, [value, setInputValue])

    return (
      <SearchBarStyle {...otherProps} {...{ hideAt, showAt }}>
        <InputContext.Provider
          value={{
            onBlur,
            onFocus,
            onChange: handleOnChange,
            placeholder,
            ...inputProps,
            ...otherProps,
          }}
        >
          <TextField
            srOnly
            keepFocus
            blurOnEscape
            onClear={handelOnClear}
            aria-label={label}
            id={inputProps?.id}
            value={inputValue}
            autoFocus={autoFocus}
            {...{
              inputProps,
              ref: inputRef,
              label,
            }}
          />
        </InputContext.Provider>
        <Button
          aria-label={placeholder || 'Zoek'}
          title={placeholder || 'Zoek'}
          type="submit"
          size={36}
          onClick={handleOnSubmit}
          variant="secondary"
        >
          <Icon size={20}>
            <div />
          </Icon>
        </Button>
        {children}
      </SearchBarStyle>
    )
  },
)

SearchBar.defaultProps = {
  placeholder: 'Search...',
  value: '',
}

export default SearchBar
