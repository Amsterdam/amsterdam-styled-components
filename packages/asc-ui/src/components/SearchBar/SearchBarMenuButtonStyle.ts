import styled from '@datapunt/asc-core'
import { color, focusStyle } from '../../utils'

interface Props {
  open: boolean
}

const SearchBarMenuButtonStyle = styled.button<Props>`
  ${focusStyle()}
  background-color: ${({ open, theme }) =>
    open
      ? color('tint', 'level5')({ theme })
      : color('tint', 'level1')({ theme })};
`

export default SearchBarMenuButtonStyle
