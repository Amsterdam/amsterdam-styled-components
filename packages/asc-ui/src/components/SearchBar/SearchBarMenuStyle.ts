import styled from '@datapunt/asc-core'
import IconButtonStyle from '../IconButton/IconButtonStyle'
import { color } from '../../utils'
import SearchBarStyle from './SearchBarStyle'

export type SearchBarMenuStyleProps = {
  open: boolean
}

const SearchBarMenuStyle = styled.div<SearchBarMenuStyleProps>`
  display: flex;
  align-items: flex-start;
  flex-direction: column;

  [aria-hidden='true'] {
    display: none;
  }

  & > ${SearchBarStyle} {
    position: relative;
    padding: 15px;
    width: 100%;
    background-color: ${color('tint', 'level2')};
  }

  & > ${IconButtonStyle} {
    height: 50px;
    width: 50px;
    background-color: ${({ open, theme }) =>
      color('tint', open ? 'level2' : 'level1')({ theme })};
  }
`

export default SearchBarMenuStyle
