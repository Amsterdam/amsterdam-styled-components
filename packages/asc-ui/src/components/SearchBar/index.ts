import SearchBarStyle from './SearchBarStyle'

export { default } from './SearchBar'

export const SearchBarStyles = {
  SearchBarStyle,
}
