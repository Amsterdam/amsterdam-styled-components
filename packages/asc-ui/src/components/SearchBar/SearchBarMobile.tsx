import React from 'react'
import Icons from '@datapunt/asc-icons'
import InputStyle from '../Input/InputStyle'
import IconButtonStyle from '../IconButton/IconButtonStyle'
import SearchBar from './SearchBar'
import ReactIcon from '../ReactIcon/Icon'
import { MenuItem, Menu } from '../..'
import SearchBarStyle from './SearchBarStyle'

ReactIcon.add(Icons.Search)

interface SearchBarProps {
  styledComponent?: any
  placeholder?: string
  label?: string
  onTextChanged: Function
  onSearch: Function
  onBlur?: Function
  onFocus?: Function
  onKeyDown?: Function
  text?: string
}

const SearchBarMobile: React.FC<SearchBarProps> = ({
  children,
  styledComponent,
  ...otherProps
}) => {
  return (
    <Menu icon={<ReactIcon type={Icons.Search} />} mobile>
      <MenuItem>
        <SearchBar styledComponent={styledComponent} {...otherProps}>
          {children}
        </SearchBar>
      </MenuItem>
    </Menu>
  )
}

SearchBarMobile.defaultProps = {
  styledComponent: SearchBarStyle,
  placeholder: 'Search...',
  onBlur: () => {},
  onFocus: () => {},
  onKeyDown: () => {},
  text: '',
}

export default SearchBarMobile
export { SearchBarStyle, InputStyle, IconButtonStyle }
