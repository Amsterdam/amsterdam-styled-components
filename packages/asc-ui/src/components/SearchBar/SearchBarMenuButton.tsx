import React from 'react'
import SearchBarMenuButtonStyle from './SearchBarMenuButtonStyle'

const SearchBarMenuButton = ({ open, ...otherProps }: any) => {
  return <SearchBarMenuButtonStyle {...{ open }} {...otherProps} />
}
export default SearchBarMenuButton
