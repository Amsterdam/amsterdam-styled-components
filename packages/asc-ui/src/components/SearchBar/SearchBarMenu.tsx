import React from 'react'
import Icons from '@datapunt/asc-assets'
import ownerDocument from '../../utils/ownerDocument'
import SearchBarMenuStyle from './SearchBarMenuStyle'
import SearchBar from './SearchBar'
import IconButton from '../IconButton/IconButton'
import ReactIcon from '../ReactIcon/Icon'
import { InputMethods } from '../Input'
import useActionOnEscape from '../../utils/useActionOnEscape'

interface SearchBarMenuProps extends InputMethods {
  css?: any
  placeholder?: string
  label?: string
  onSubmit?: Function
}

interface Props extends SearchBarMenuProps {
  open?: boolean
}
const SearchBarMenu: React.FC<Props> = ({ children, css, ...otherProps }) => {
  const ref = React.useRef<HTMLDivElement>(null)
  const [open, setOpen] = React.useState(false)
  const { onKeyDown } = useActionOnEscape(() => setOpen(false))

  const onClose = () => {
    setTimeout(() => {
      const element = ref.current as HTMLInputElement
      if (element) {
        const currentFocus = ownerDocument(element).activeElement
        if (!element.contains(currentFocus)) {
          setOpen(false)
        }
      }
    })
  }

  return (
    <SearchBarMenuStyle
      {...{
        onKeyDown,
        css,
        ref,
        open,
      }}
      onBlur={onClose}
    >
      <IconButton
        aria-label="Search"
        onClick={() => {
          setOpen(!open)
        }}
        iconSize={open ? 16 : 20}
      >
        <ReactIcon type={open ? Icons.Close : Icons.Search} />
      </IconButton>

      {open && <SearchBar {...otherProps}>{children}</SearchBar>}
    </SearchBarMenuStyle>
  )
}

SearchBarMenu.defaultProps = {
  css: '',
  placeholder: 'Search...',
}

export default SearchBarMenu
