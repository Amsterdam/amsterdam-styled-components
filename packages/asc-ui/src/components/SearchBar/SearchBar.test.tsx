import { render, fireEvent } from '@testing-library/react'
import SearchBar from './SearchBar'

jest.mock('polished')

describe('SearchBar', () => {
  let container: any

  const mockOnSubmit = jest.fn()

  beforeEach(() => {
    ;({ container } = render(
      <SearchBar onSubmit={mockOnSubmit}>searchbar-content</SearchBar>,
    ))
  })

  it('should render the label and icon', () => {
    const icon = container.querySelector('button span')

    expect(icon).toBeDefined()
    expect(container.firstChild).toHaveTextContent('searchbar-content')
  })

  it('should handle the onClick event', () => {
    const button = container.querySelector('button')

    fireEvent.click(button)

    expect(mockOnSubmit).toHaveBeenCalled()
  })
})
