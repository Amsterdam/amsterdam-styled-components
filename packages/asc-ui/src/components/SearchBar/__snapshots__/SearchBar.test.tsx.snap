// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render 1`] = `
.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 20px;
  height: 20px;
  padding: undefinedpx;
  box-sizing: content-box;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c7 > svg {
  width: inherit;
  height: inherit;
}

.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: none;
  cursor: pointer;
  font-size: 16px;
  font-weight: normal;
  padding: 0 10px 0 10px;
  -webkit-transition: color 0.1s ease-in-out,background-color 0.1s ease-in-out;
  transition: color 0.1s ease-in-out,background-color 0.1s ease-in-out;
  background: #ec0000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 5px;
  width: 30px;
  height: 30px;
}

.c6:focus {
  outline-color: #FEC813;
  outline-style: solid;
  outline-width: 3px;
}

.c6:hover {
  background: #bc0000;
}

.c6:after {
  content: '';
  display: block;
  min-height: inherit;
  font-size: 0;
}

.c4 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-size: 1rem;
  border: solid 1px #767676;
  border-radius: 0;
  box-sizing: border-box;
  padding: 10px 10px;
  width: 100%;
}

.c3 {
  border-width: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c2 {
  position: relative;
}

.c2 > .c5 {
  position: absolute;
  top: 5px;
  right: 5px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 .c1 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c0 > .c5 {
  margin-left: 5px;
  padding: 10px;
  width: 40px;
  height: 40px;
}

.c0 > .c5 svg rect,
.c0 > .c5 svg polygon,
.c0 > .c5 svg path {
  fill: #ffffff;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <label
      class="c3"
      for="search-bar-id"
      type="text"
    >
      Search...
    </label>
    <input
      aria-label="Search..."
      class="c4"
      id="search-bar-id"
      placeholder="Search..."
      type="text"
      value=""
    />
  </div>
  <button
    aria-label="Search"
    class="c5 c6"
    color="secondary"
  >
    <span
      class="c7"
    >
      <svg-mock />
    </span>
  </button>
  searchbar-content
</div>
`;
